oscore-cf/src/main/java/org/eclipse/californium/oscore/ObjectSecurityLayer.java

```diff
@@ -0,0 +1,297 @@
+				Response error;
+				try {
+					error = CoapOSExceptionHandler.manageError(e, request);
+					if (error != null) {
+						super.sendResponse(exchange, error);
+					}
+					return;
+				} catch (OSException e1) {
+					e1.printStackTrace();
+				}
+			}
+			exchange.setCryptographicContextID(rid);
+		}
+		super.receiveRequest(exchange, request);
+	}
+
+	@Override
+	public void receiveResponse(Exchange exchange, Response response) {
+		Request request = exchange.getCurrentRequest();
+		if (request == null) {
+			LOGGER.error("No request tied to this response");
+			return;
+		}
+		try {
+			if (responseShouldBeProtected(exchange, response)) {
+				if (!isProtected(response)) {
+					LOGGER.error("Unprotected OSCore response received" + " for OSCore request");
+					LOGGER.error("Response: " + response.toString());
+					return;
+				}
+				response = prepareReceive(response);
+
+			}
+		} catch (OSException e) {
+			LOGGER.error("Error while receiving OSCore response: " + e.getMessage());
+			EmptyMessage error = CoapOSExceptionHandler.manageError(e, response);
+			if (error != null) {
+				sendEmptyMessage(exchange, error);
+			}
+			return;
+		}
+		super.receiveResponse(exchange, response);
+	}
+
+	@Override
+	public void receiveEmptyMessage(Exchange exchange, EmptyMessage message) {
+		super.receiveEmptyMessage(exchange, message);
+	}
+
+	private static boolean shouldProtectResponse(Exchange exchange) {
+		return exchange.getCryptographicContextID() != null;
+	}
+
+	private static boolean responseShouldBeProtected(Exchange exchange, Response response) throws OSException {
+		Request request = exchange.getCurrentRequest();
+		OptionSet options = request.getOptions();
+		if (exchange.getCryptographicContextID() == null) {
+			if (response.getOptions().hasObserve() && request.getOptions().hasObserve()) {
+
+				// Since the exchange object has been re-created the
+				// cryptographic id doesn't exist
+				if (options.hasOption(OptionNumberRegistry.OSCORE)) {
+					String uri = request.getURI();
+					OSCoreCtxDB db = HashMapCtxDB.getInstance();
+					OSCoreCtx ctx = null;
+					try {
+						ctx = db.getContext(uri);
+					} catch (OSException e) {
+						LOGGER.error("Error when re-creating exchange at OSCORE level");
+						throw new OSException("Error when re-creating exchange at OSCORE level");
+					}
+					exchange.setCryptographicContextID(ctx.getRecipientId());
+				}
+			}
+		}
+		return exchange.getCryptographicContextID() != null;
+	}
+
+	private static boolean shouldProtectRequest(Request request) {
+		OptionSet options = request.getOptions();
+		return options.hasOption(OptionNumberRegistry.OSCORE);
+
+	}
+
+	private static boolean isProtected(Message message) {
+		return message.getOptions().getOscore() != null;
+	}
+
+	/**
+	 *
+	 * Exists to provide better adding of ctx and seqByToken when the outgoing
+	 * request has a valid Token.
+	 *
+	 */
+	private class OSCoreMessageObserver extends MessageObserverAdapter {
```

Comment:
I think, a anonymous inner class will be less overkill.

```
				String uri = request.getURI();
				final OSCoreCtxDB db = HashMapCtxDB.getInstance();

				if (uri == null) {
					LOGGER.error(Error.URI_NULL);
					throw new OSException(Error.URI_NULL);
				}
				if (db == null) {
					LOGGER.error(Error.DB_NULL);
					throw new OSException(Error.DB_NULL);
				}

				final OSCoreCtx ctx = db.getContext(uri);
				if (ctx == null) {
					LOGGER.error(Error.CTX_NULL);
					throw new OSException(Error.CTX_NULL);
				}

				exchange.setCryptographicContextID(ctx.getRecipientId());
				final int seqByToken = ctx.getSenderSeq();

				final Request preparedRequest = prepareSend(request, ctx);
				preparedRequest.addMessageObserver(new MessageObserverAdapter() {
					@Override
					public void onReadyToSend() {
						try {
							Token token = preparedRequest.getToken();
							db.addContext(token, ctx);
							db.addSeqByToken(token, seqByToken);
						} catch (OSException e) {
							LOGGER.error(Error.TOKEN_NULL);
							throw new IllegalArgumentException(Error.TOKEN_NULL);
						}
					}
				});
				req = preparedRequest;
```

Just ensure, that all variables accessed in the callback are final.